pkg/agent/client.go

```diff
@@ -5,50 +5,116 @@ import (
+	caContentProvider *dynamiccertificates.ConfigMapCAController
+}
+
+var _ dynamiccertificates.Listener = &antreaClientProvider{}
+
+func NewAntreaClientProvider(config config.ClientConnectionConfiguration, kubeClient kubernetes.Interface) *antreaClientProvider {
+	antreaCAProvider, _ := dynamiccertificates.NewDynamicCAFromConfigMapController("antrea-ca", cert.CAConfigMapNamespace, cert.CAConfigMapName, cert.CAConfigMapKey, kubeClient)
+	antreaClientProvider := &antreaClientProvider{
+		config:            config,
+		caContentProvider: antreaCAProvider,
+	}
+
+	antreaCAProvider.AddListener(antreaClientProvider)
+	return antreaClientProvider
+}
+
+// RunOnce runs the task a single time synchronously, ensuring client is initialized.
+func (p *antreaClientProvider) RunOnce() error {
+	p.caContentProvider.RunOnce()
+	return p.updateAntreaClient()
+}
+
+// Run starts the caContentProvider, which watches the ConfigMap and notifies changes
+// by calling Enqueue.
+func (p *antreaClientProvider) Run(stopCh <-chan struct{}) {
+	p.caContentProvider.Run(1, stopCh)
+}
+
+// Enqueue implements dynamiccertificates.Listener. It will be called by caContentProvider
+// when caBundle is updated.
+func (p *antreaClientProvider) Enqueue() {
+	if err := p.updateAntreaClient(); err != nil {
+		klog.Errorf("Failed to update Antrea client: %v", err)
+	}
+}
+
+// GetAntreaClient implements AntreaClientProvider.
+func (p *antreaClientProvider) GetAntreaClient() (versioned.Interface, error) {
+	p.mutex.RLock()
+	defer p.mutex.RUnlock()
+	return p.client, nil
+}
+
+func (p *antreaClientProvider) updateAntreaClient() error {
+	caBundle := p.caContentProvider.CurrentCABundleContent()
+	if caBundle == nil {
+		klog.Warningf("Didn't get CA certificate from ConfigMap, may not be able to verify server cert")
+	}
 	var kubeConfig *rest.Config
 	var err error
-
-	if len(config.Kubeconfig) == 0 {
+	if len(p.config.Kubeconfig) == 0 {
 		klog.Info("No antrea kubeconfig file was specified. Falling back to in-cluster config")
-		kubeConfig, err = inClusterConfig()
+		kubeConfig, err = inClusterConfig(caBundle)
 	} else {
 		kubeConfig, err = clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
-			&clientcmd.ClientConfigLoadingRules{ExplicitPath: config.Kubeconfig},
+			&clientcmd.ClientConfigLoadingRules{ExplicitPath: p.config.Kubeconfig},
 			&clientcmd.ConfigOverrides{}).ClientConfig()
 	}
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	// ContentType will be used to define the Accept header if AcceptContentTypes is not set.
 	kubeConfig.ContentType = "application/vnd.kubernetes.protobuf"
-	kubeConfig.QPS = config.QPS
-	kubeConfig.Burst = int(config.Burst)
-
+	kubeConfig.QPS = p.config.QPS
+	kubeConfig.Burst = int(p.config.Burst)
 	client, err := versioned.NewForConfig(kubeConfig)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	return client, nil
+	klog.Info("Updating Antrea client with the new CA bundle")
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
```

Comment:
Just wonder if locking is really needed in Go to update a pointer?

Reply:
Yes, it needs locking: https://golang.org/ref/mem#tmp_10.
Run this code with race detector:
```
func TestRace(t *testing.T) {
	a := &A{a: "a1"}
	go func() {
		b := a
		t.Log(b)
	}()
	go func() {
		a = &A{a: "a2"}
	}()
}
```
Will result in:
```
==================
WARNING: DATA RACE
Write at 0x00c0003506c8 by goroutine 70:
  github.com/vmware-tanzu/antrea/pkg/apiserver/certificate.TestRace.func2()
      /root/antrea/pkg/apiserver/certificate/certificate_test.go:189 +0x7f

Previous read at 0x00c0003506c8 by goroutine 69:
  github.com/vmware-tanzu/antrea/pkg/apiserver/certificate.TestRace.func1()
      /root/antrea/pkg/apiserver/certificate/certificate_test.go:185 +0x3c

Goroutine 70 (running) created at:
  github.com/vmware-tanzu/antrea/pkg/apiserver/certificate.TestRace()
      /root/antrea/pkg/apiserver/certificate/certificate_test.go:188 +0xfc
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199

Goroutine 69 (finished) created at:
  github.com/vmware-tanzu/antrea/pkg/apiserver/certificate.TestRace()
      /root/antrea/pkg/apiserver/certificate/certificate_test.go:184 +0xda
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:909 +0x199
==================
```

Reply:
Ok

Reply:
But will it create real issues in our case? We do not care the old or new value is returned right? The question is whether updating pointer itself is atomic.

Reply:
I think there is no guarantee in Go that a pointer update is atomic, even though it may be the case on most 64-bit systems. So if another goroutine accesses `p.client` while we write to it, that's a race.